File: tests/_integration/scell_seal_string_echo.go
package main

import (
	"encoding/base64"
	"fmt"
	"github.com/cossacklabs/themis/gothemis/cell"
	"os"
)

func main() {
	args := os.Args

	if len(args) < 4 || len(args) > 5 {
		fmt.Printf("usage %s <command: enc | dec > <key> <message> <context>\n", os.Args[0])
		os.Exit(1)
	}

	command := args[1]
	key := args[2]
	message := args[3]
	var context []byte

	if len(args) == 5 {
		context = []byte(args[4])
	}

	sc := cell.New([]byte(key), cell.CELL_MODE_SEAL)

	if "enc" == command {
		encData, _, err := sc.Protect([]byte(message), context)
		if nil != err {
			fmt.Println(os.Stderr, "Error encrypting message")
			os.Exit(1)
		}
		fmt.Println(base64.StdEncoding.EncodeToString(encData))

	} else if "dec" == command {
		decoded_message, err := base64.StdEncoding.DecodeString(message)
		if nil != err {
			fmt.Println(os.Stderr, "Error decoding message")
			os.Exit(1)
		}
		decData, err := sc.Unprotect(decoded_message, nil, context)
		if nil != err {
			fmt.Println(os.Stderr, "Error decrypting message")
			os.Exit(1)
		}
		fmt.Println(string(decData[:]))

	} else {
		fmt.Println(os.Stderr, "Wrong command, use \"enc\" or \"dec\"")
		os.Exit(1)
	}
}

File: tools/js/wasm-themis/scell_seal_string_echo.js
const themis = require('wasm-themis')

let command, key, message, context
if (5 <= process.argv.length && process.argv.length <= 6) {
    command = process.argv[2]
    key     = process.argv[3]
    message = process.argv[4]
    if (process.argv.length == 6) {
        context = process.argv[5]
    }
} else {
    console.log('usage: node scell_seal_string_echo.js <enc|dec> <key> <message> [<context>]')
    process.exit(1);
}

themis.initialized.then(function() {
    let cell = themis.SecureCellSeal.withKey(Buffer.from(key))
    let result
    switch (command) {
        case 'enc':
            if (context) {
                result = cell.encrypt(Buffer.from(message), Buffer.from(context))
            } else {
                result = cell.encrypt(Buffer.from(message))
            }
            console.log(Buffer.from(result).toString('base64'))
            break
        case 'dec':
            if (context) {
                result = cell.decrypt(Buffer.from(message, 'base64'), Buffer.from(context))
            } else {
                result = cell.decrypt(Buffer.from(message, 'base64'))
            }
            console.log(Buffer.from(result).toString('utf-8'))
            break
        default:
            console.log('invalid command "' + command + '": use "enc" or "dec"')
            process.exit(1)
    }
})
